src/server/mod.rs

```diff
@@ -1812,8 +1812,152 @@ impl LanguageServer for LspServer {
 
                 Ok(None)
             }
-            Ok(LspServerCommand::AddFieldFilter) => todo!(),
-            Ok(LspServerCommand::RemoveFieldFilter) => todo!(),
+            Ok(LspServerCommand::AddFieldFilter) => {
+                let command_params: ValueFilterParams =
+                    match serde_json::value::from_value(
+                        params.arguments[0].clone(),
+                    ) {
+                        Ok(value) => value,
+                        Err(err) => {
+                            return Err(LspError::InternalError(
+                                format!("{:?}", err),
+                            )
+                            .into())
+                        }
+                    };
+
+                let file = self.store.get_ast_file(
+                    &command_params.text_document.uri,
+                )?;
+                let mut composition =
+                    composition::Composition::new(file);
+                let status =
+                    composition.add_field(command_params.value);
+                if status.is_err() {
+                    return Err(LspError::InternalError(
+                        "Failed to add field to composition."
+                            .to_string(),
+                    )
+                    .into());
+                }
+                let new_text = composition.to_string();
+
+                let last_pos =
+                    line_col::LineColLookup::new(&new_text)
+                        .get(new_text.len());
+
+                let edit = lsp::WorkspaceEdit {
+                    changes: Some(HashMap::from([(
+                        command_params.text_document.uri.clone(),
+                        vec![lsp::TextEdit {
+                            new_text: new_text.clone(),
```

Comment:
is this wiping out the entire query to update it with the complete modified query? does that cross the wasm / browser interface? if it does that'd be real slow for large queries.

Reply:
Excellent question.

It's not doing a replacement of all the text in the editor.
Instead, it replaces just the composition block.

A "composition" is defined as a single subquery ending in the proper yield identifier; `from(bucket) |> filter() |> filter |> yield(uniqueIdForComposition)`. Schema composition is not append only; instead, it can modify any part within this single subquery it's composing. (A.k.a. an active schema composition has the LSP "owning" the composition subquery.)

Therefore, each time the composition block is edited by the LSP; we do a replacement of only the composition block (the composition subquery). The rest of the editor text remains untouched.

Reply:
gotcha. so I am reading the "replacing the whole query every time part" correctly, but you're saying that the size of that mutation is scoped to the workspace which should remain limited

Reply:
exactly.

Although an fyi for the terminology used:
* workspace == multiple files.
   * LSP spec definition.
   * e.g. our variable definitions are sent to the LSP in a separate file, within the same workspace.
* file:
   * LSP spec definition.
   * a.k.a. editor text is a single file.
* my terminology "subquery" == a subset of lines, in a single code block, within a single File.
  * a single File can have multiple subqueries.
  * a subquery is the `from(src) |> transformations |> yield()`
* the "composition block" == a specific subquery
   * A "composition" is defined as a single subquery ending in the proper yield identifier
   * `from(bucket) |> filter() |> filter |> yield(uniqueIdForComposition)`